substrate/client/network/sync/src/service/syncing_service.rs

```diff
@@ -91,6 +101,14 @@ impl<B: BlockT> SyncingService<B> {
 		rx.await
 	}
 
+	/// Restart the synchronization with new arguments.
+	pub async fn restart(&self, sync_restart_args: SyncRestartArgs<B>) {
```

Comment:
If there is no fast sync anymore, do we still need to restart sync? If so then why?

Reply:
We'll start the fast-sync using `SyncMode::Light` configuration to allow the block insertion without the sequence check. Later we need to switch to the `SyncMode::Full` to download the full blocks on a regular sync. As I mentioned in the description, Substrate sync can switch to the full mode automatically but it will generate weird messages. It is a UX improvement.

Reply:
I guess I don't understand why `SyncMode::Light` is necessary in the first place. We don't want to sync, we just want to insert one block with its state into the database.

Reply:
The exact error is `InvalidBlockNumber` within `NonCanonicalOverlay` struct (`insert` method). It seems to trigger when `LAST_CANONICAL` db meta entry is set which in turn seems to be set when we invoke `try_commit_operation` (canonicalize_block() when operation.commit_state is true). Setting `LightState` sets `commit_state` initially to false when we invoke `set_genesis_state` having 
```
/// Returns true if the genesis state writing will be skipped while initializing the genesis
/// block.
pub fn no_genesis(&self) -> bool {
   matches!(self.network.sync_mode, SyncMode::LightState { .. } | SyncMode::Warp { .. })
}
```

This way we disable overlay related checks when we initially don't commit state.

Reply:
And how do we achieve the same without light state sync?

Reply:
Not sure I understood the question. Without the light state sync we hit the `InvalildBlockNumber` error trying to fast sync.

Reply:
And I'm saying we need to create an API that doesn't hit that error when we try to insert a block

Reply:
It's likely possible by hijacking `lock_import_and_run` method and parameterizing `commit_state` variable but it would be very invasive within the current Substrate architecture. Switching the sync mode before passing the control is much less "foreign" than the proposed change if I understand you correctly.

Reply:
Is that what upstream suggested?

Reply:
No. I did this call stack research based on your Thursday's request.

Reply:
To be completely honest and transparent I'm getting annoyed and frustrated that it takes forever to convince you to talk to upstream :confused: